examples/tensorflow/text-classification/run_glue.py

```diff
@@ -456,13 +417,13 @@ def compute_metrics(preds, label_ids):
             dataset = datasets[key]
             if samples_limit is not None:
                 dataset = dataset.select(range(samples_limit))
-            data = convert_dataset_for_tensorflow(
-                dataset,
-                non_label_column_names,
+            data = dataset.to_tf_dataset(
+                columns=[col for col in dataset.column_names if col not in set(non_label_column_names + ["label"])],
+                shuffle=shuffle,
                 batch_size=batch_size,
-                dataset_mode=dataset_mode,
+                collate_fn=data_collator,
                 drop_remainder=drop_remainder,
-                shuffle=shuffle,
+                label_cols="label" if "label" in dataset.column_names else None,
```

Comment:
From the [docs](https://huggingface.co/docs/datasets/master/package_reference/main_classes.html#datasets.Dataset.to_tf_dataset) in `to_tf_dataset`, I would think that not setting `label_cols` would be fine for training purposes. However, training requires this field and also fails if `labels` are part of the input. Double-checking: it is the right behavior, correct?

If yes, I'm adding a PR to clear out the docs (that setting `label_cols` is required to fit). Maybe also an assertion that the label col can't be in the input cols? I'd say is a recipe for potential data leakage, having the labels in the inputs.

Reply:
cc @Rocketknight1 for the internal of `to_tf_datatset`

Reply:
This is a good question, and I should probably document it more clearly (including in the `to_tf_dataset` docs). `tf.data.Dataset` objects emit either a dict `features` or a tuple of two dicts `(features, labels)` when you iterate over them. The `columns` argument sets which columns of the input dataset will be used as keys in the `features` dict and the `label_cols` argument sets which columns of the input dataset will be used as keys in the `labels` dict. If `label_cols` is not set then the `tf.data.Dataset` will only emit a `features` dict.

The default behaviour of Keras is that it expects to get `(features, labels)` tuples as input. It does a forward pass with the `features` to get model outputs, and then passes the outputs and labels to the loss function specified in `compile()` to get a loss.

The problem with that, though, is that our models often have weird losses that are really painful for users to implement correctly - e.g. BERT has a token cross-entropy loss which should only be computed on tokens selected for masking, plus a next-sentence prediction loss. No user is ever going to correctly specify a Keras loss function for that, so we need to help them out.

The solution I used was to take advantage of the fact that our models compute a `loss` tensor during the forward pass if you pass the required labels as keys in the input dict. This is the standard way to train them with PyTorch - you just pass everything in the input dict and then do `outputs.loss.backward()`. In Keras, you can do the same thing - you just specify a "dummy" loss function that reads that loss value and outputs it directly as the loss.

So if the user `compile`s our models without a `loss` argument, we interpret that as them wanting to use the internal loss, and we quietly stick a dummy loss function on the `loss` output head of the model. If they're doing this, they have to include all the label columns as keys in the input dict (`compile()` throws a warning when you don't specify a loss to remind you to do this). If, on the other hand, the user specifies a loss when calling `compile()`, then we just use the standard Keras behaviour. In that case, it doesn't matter whether the label keys are in the input dict or not because we totally ignore the `loss` output head, but the user does have to pass a separate labels array or dict (usually by using the `label_cols` argument to `to_tf_dataset`) so that Keras can find them.

Reply:
Amen to all of the above ;-)

Reply:
@Rocketknight1 added a few more comments to attempt to summarily explain what you wrote above :)